plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts

```diff
@@ -257,14 +260,14 @@ function computeKey(pluginEvent: PipelineEvent): string {
     return `${pluginEvent.team_id ?? pluginEvent.token}:${pluginEvent.distinct_id}`
 }
 
-async function emitToOverflow(queue: IngestionConsumer, kafkaMessages: Message[]) {
+async function emitToOverflow(queue: IngestionConsumer, kafkaMessages: Message[], useRandomPartitioner: boolean) {
     ingestionOverflowingMessagesTotal.inc(kafkaMessages.length)
     await Promise.all(
         kafkaMessages.map((message) =>
             queue.pluginsServer.kafkaProducer.produce({
                 topic: KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW,
                 value: message.value,
-                key: null, // No locality guarantees in overflow
+                key: useRandomPartitioner ? undefined : message.key,
```

Comment:
there's a change here where we set it to `undefined` instead of `null` and later we check for `null`

Reply:
As far as I can tell, this was the issue.

The `produce` call here [eventually forwards to a `HighLevelProducer` instance](https://github.com/PostHog/posthog/blob/30bafdd38278b3fb6d20f53984df3b0154cf4a37/plugin-server/src/utils/db/kafka-producer-wrapper.ts#L42-L52), which [does some extra stuff with the `key` in produce](https://github.com/Blizzard/node-rdkafka/blob/0b5987aeecbea413334d7b7ad04b90a10eb006c8/lib/producer/high-level-producer.js#L276-L290). If the `key` ends up being `undefined` it [just quietly doesn't produce the message](https://github.com/Blizzard/node-rdkafka/blob/0b5987aeecbea413334d7b7ad04b90a10eb006c8/lib/producer/high-level-producer.js#L207-L209), and the callback is never called.

This is also the behavior that I saw testing this manually in the node REPL:

```node
> const { HighLevelProducer } = require('node-rdkafka')
undefined
> const p = new HighLevelProducer({'bootstrap.servers': 'kafka:9092'})
undefined
> p.connect()
// truncated
> > p.produce('garbage-topic', null, Buffer.from('message'), 'key', undefined, (...a) => { console.log('callback:', a) })
undefined
> callback: [ null, 1 ]
> p.produce('garbage-topic', null, Buffer.from('message'), undefined, undefined, (...a) => { console.log('callback:', a) })
undefined
// nothing ever happens here
```

What I figure happened is that any message that should have been routed to overflow never resolved it's promise, and the consumers simply stopped making forward progress once they saw a batch containing one of those messages.

[The `key` property is typed as `MessageKey`](https://github.com/PostHog/posthog/blob/30bafdd38278b3fb6d20f53984df3b0154cf4a37/plugin-server/src/utils/db/kafka-producer-wrapper.ts#L35), which [does include undefined](https://github.com/Blizzard/node-rdkafka/blob/e7e4c6d1acbd3fa508b50a57133e2fa60d3903ac/index.d.ts#L95), and the [`HighLevelProducer.produce` signature accepts `any`](https://github.com/Blizzard/node-rdkafka/blob/e7e4c6d1acbd3fa508b50a57133e2fa60d3903ac/index.d.ts#L281-L282), which explains why this wasn't caught by the type checker.